### PR TITLE
Partially implements GameWindow.

### PR DESCRIPTION
**Pull request by tomsons26**

### CODE REVIEW

File: src/game/client/gui/gamewindow.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+enum WindowMsgHandledType
+{
+    MSG_IGNORED,
+    MSG_HANDLED,
+};
+
+typedef WindowMsgHandledType(__cdecl *window_callback_func)(
+    GameWindow *window, unsigned int message, unsigned int data_1, unsigned int data_2);
+
+typedef void(__cdecl *window_draw_func)(GameWindow *window, WinInstanceData *instance);
```
Comment (duncanspumpkin) [inline src/game/client/gui/gamewindow.h]: I don't think there is an entry in the style guide for function pointers but I would expect them to follow either the casing of classes or the casing of functions and not the style of variables.

File: src/game/client/gui/gamewindow.cpp [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    if (m_inputData != nullptr) {
+        delete m_inputData;
+    }
+    m_inputData = nullptr;
+
+    if (m_editData != nullptr) {
+        delete m_editData;
+    }
+
+    m_editData = nullptr;
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: Destructor can be simplified to just
```
delete m_inputData;
delete m_editData;
```

File: src/game/client/gui/gamewindow.cpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+}
+
+void GameWindow::Normalize_Window_Region()
+{
+    if (m_region.lo.x > m_region.hi.x) {
+        int tmp = m_region.lo.x;
+        m_region.lo.x = m_region.hi.x;
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: Can be simplified to `std::swap(m_region.lo.x, m_region.hi.x)`

File: src/game/client/gui/gamewindow.cpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        m_region.hi.x = tmp;
+    }
+
+    if (m_region.lo.y > m_region.hi.y) {
+        int tmp = m_region.lo.y;
+        m_region.lo.y = m_region.hi.y;
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: Can be simplified to `std::swap(m_region.lo.y, m_region.hi.y)`

File: src/game/client/gui/gamewindow.cpp [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+
+    Normalize_Window_Region();
+
+    return 0;
+}
+
+int GameWindow::Win_Get_Position(int *x, int *y)
+{
+    if (!x || !y) {
+        return -3;
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp:81]: What does this mysterious -3 mean?

File: src/game/client/gui/gamewindow.cpp [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+
+    return 0;
+}
+
+int GameWindow::Win_Get_Screen_Position(int *x, int *y)
+{
+    GameWindow *parent = m_parent;
+
+    *x = m_region.lo.x;
+    *y = m_region.lo.y;
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp:101]: In above function, `x` and `y` are tested, but here not. Is this proper?

File: src/game/client/gui/gamewindow.cpp [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+
+    *width = m_size.x;
+    *height = m_size.y;
+
+    return 0;
+}
+
+int GameWindow::Win_Enable(bool enable)
+{
+    int status;
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: This temporary can be removed. Just work on `m_status` directly.
```c++
    if (enable) {
        m_status |= WIN_STATUS_ENABLED;
    } else {
        m_status &= ~WIN_STATUS_ENABLED;
    }
```

File: src/game/client/gui/gamewindow.cpp [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+        int r_x = i->m_region.lo.x;
+        int r_y = i->m_region.lo.y;
+
+        for (GameWindow *j = i->m_parent; j != nullptr; j = j->m_parent) {
+            r_x += j->m_region.lo.x;
+            r_y += j->m_region.lo.y;
+        }
+
+        if (x >= r_x && x <= i->m_size.x + r_x && y >= r_y && y <= i->m_size.y + r_y
+            && (ignore_hidden != 1 || !(i->m_status & WIN_STATUS_HIDDEN))) {
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: `!ignore_hidden` or `ignore_hidden == false`

File: src/game/client/gui/gamewindow.cpp [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+        }
+    }
+
+    return this;
+}
+
+int GameWindow::Win_Set_Enabled_Image(int index, Image *image)
+{
+    if (index < 0 || index >= WinInstanceData::MAX_DRAW_DATA) {
+        captainslog_debug("Set enabled image, index out of range '%d'\n", index);
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: `\n` char at the end can be omitted in all logs.

File: src/game/client/gui/gamewindow.cpp [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+int GameWindow::Win_Set_Tooltip_Func(window_tooltip_func tooltip)
+{
+    m_tooltip = tooltip;
+
+    return 0;
+}
+
+GameWindow *GameWindow::Win_Point_In_Child(int x, int y, bool ignore_enable_check, bool audio_event)
+{
+    for (GameWindow *i = m_child; i != nullptr; i = i->m_next) {
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: i = window

File: src/game/client/gui/gamewindow.cpp [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+        int r_x = i->m_region.lo.x;
+        int r_y = i->m_region.lo.y;
+
+        for (GameWindow *j = i->Win_Get_Parent(); j != nullptr; j = j->m_parent) {
```
Comment (xezon) [inline src/game/client/gui/gamewindow.cpp]: j = parent_window

Same in next function.

File: src/game/client/gui/gamewindow.h (hunk 1 above)
Comment (OmniBlade) [reply src/game/client/gui/gamewindow.h]: I'm not sure if I've done any function pointers as of yet, I've probably done it as one of the type styles, callback_func_t or CallbackFunc. I might be inclined to drop the "func" when callback is also mentioned as well, callback implies it. So shall we go with WindowCallback and WindowDrawFunc?

**Pull request merged**